src/index.ts

```diff
@@ -98,22 +121,72 @@ const memoryOnlyPrecomputedBanditsStore = precomputedBanditStoreFactory();
  * @public
  */
 export class EppoJSClient extends EppoClient {
-  // Ensure that the client is instantiated during class loading.
-  // Use an empty memory-only configuration store until the `init` method is called,
-  // to avoid serving stale data to the user.
+  /**
+   * Resolved when the client is initialized
+   * @private
+   */
+  private readonly readyPromise: Promise<void>;
+
+  /**
+   * Used when the client is initialized from outside the constructor, namely, from the `init` or
+   * `EppoJSClient.initializeClient` methods.
+   * @private
+   */
+  private readyPromiseResolver: (() => void) | null = null;
+
   public static instance = new EppoJSClient({
     flagConfigurationStore,
     isObfuscated: true,
   });
-  public static initialized = false;
+  initialized = false;
+
+  constructor(optionsOrConfig: EppoClientParameters | IClientOptions) {
+    const v2Constructor = 'sdkKey' in optionsOrConfig;
+
+    super(
+      v2Constructor
+        ? clientOptionsToEppoClientParameters(
+            optionsOrConfig, // The IClientOptions wrapper
+            optionsOrConfig.flagConfigurationStore ?? // create a new, memory only FCS if none was provided
+              configurationStorageFactory({
+                forceMemoryOnly: true,
+              }),
+            optionsOrConfig.eventIngestionConfig // Create an Event Dispatcher if Event Ingestion config was provided.
+              ? newEventDispatcher(optionsOrConfig.sdkKey, optionsOrConfig.eventIngestionConfig)
+              : undefined,
+          ) // "New" construction technique; isolated instance.
+        : (optionsOrConfig as EppoClientParameters), // Legacy instantiation of singleton.
+    );
+
+    if (!v2Constructor) {
+      // Original constructor path; passthrough to super is above.
+      // `waitForReady` is a new API we need to graft onto the old way of constructing first and initializing later.
+
+      // Create a promise that will be resolved when EppoJSClient.initializeClient() is called
+      this.readyPromise = new Promise((resolve) => {
+        this.readyPromiseResolver = resolve;
+      });
+    } else {
+      this.readyPromise = explicitInit(
```

Comment:
what happens if an error is hit? I think we may need outer try-catch here to respect our throw on failed initialization flag. And if that is false (which we hope to make the new default) we'll probably want to return a stand-in client similar to what the singleton has.

Reply:
init handles catching exceptions on init and throwing/logging depending on the option